py/variable.cpp

```diff
@@ -44,10 +37,18 @@ Variable_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
    if( name != 0 )
    {
 #if PY_MAJOR_VERSION >= 3
-    	new( &self->variable ) kiwi::Variable( name );
+      if( !PyUnicode_Check( name ) )
+    	   return py_expected_type_fail( name, "unicode" );
 #else
-    	new( &self->variable ) kiwi::Variable( PyString_AS_STRING( name ) );
+      if( !(PyString_Check( name ) | PyUnicode_Check( name ) ) )
+      {
+         return py_expected_type_fail( name, "str or unicode" );
+      }
 #endif
+      std::string c_name;
+      if( !convert_pystr_to_str(name, c_name) )
+          return 0;
+    	new( &self->variable ) kiwi::Variable( c_name );
```

Comment:
Since this change causes the function to possible return early after the python variable object has been allocated, we need to use a smart pointer to guard the lifetime of that object. Here's an updated version of this function which does that:

```c++
static PyObject*
Variable_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
{
	static const char *kwlist[] = { "name", "context", 0 };
	PyObject* context = 0;
	PyObject* name = 0;

	if( !PyArg_ParseTupleAndKeywords(
		args, kwargs, "|OO:__new__", const_cast<char**>( kwlist ),
		&name, &context ) )
		return 0;

	PyObjectPtr pyvar( PyType_GenericNew( type, args, kwargs ) );
	if( !pyvar )
		return 0;

	Variable* self = reinterpret_cast<Variable*>( pyvar.get() );
	self->context = xnewref( context );

	if( name != 0 )
	{
#if PY_MAJOR_VERSION >= 3
		if( !PyUnicode_Check( name ) )
			return py_expected_type_fail( name, "unicode" );
#else
		if( !( PyString_Check( name ) | PyUnicode_Check( name ) ) )
		{
			return py_expected_type_fail( name, "str or unicode" );
		}
#endif
		std::string c_name;
		if( !convert_pystr_to_str(name, c_name) )
			return 0;
		new( &self->variable ) kiwi::Variable( c_name );
	}
	else
	{
		new( &self->variable ) kiwi::Variable();
	}

	return pyvar.release();
}
```

Reply:
Thanks for catching this.